### PR TITLE
RDS and S3 support for Kubeflow on Terraform

### PR DESCRIPTION
**Pull request by rrrkharse**
**Description of your changes:**
- Added deployment options for:
  - RDS only
  - S3 only
  - RDS and S3
  - Vanilla (but can toggle RDS and/or S3 on with a boolean flag)
- Added RDS and S3 related AWS terraform module
- Configured KFP and Katib modules to be dynamically configured with the added deployment options

**Testing:**
- Manual pipeline runs

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: deployments/rds-s3/terraform/rds-s3-components/main.tf [hunk 1]
```diff
@@ -0,0 +1,329 @@
+locals {
+  katib_chart_vanilla  = "${var.kf_helm_repo_path}/charts/apps/katib/-external-db-with-kubeflow"
```
Comment (surajkota) [inline deployments/rds-s3/terraform/rds-s3-components/main.tf:2]: incorrect paths?
```
https://github.com/awslabs/kubeflow-manifests/tree/main/charts/apps/katib/vanilla
https://github.com/awslabs/kubeflow-manifests/tree/main/charts/apps/katib/**katib**-external-db-with-kubeflow
```

File: deployments/rds-s3/terraform/rds-s3-components/main.tf [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    "rds.secretName" = try(module.rds[0].rds_secret_name, null),
+    "s3.secretName" = try(module.s3[0].s3_secret_name, null),
+  }
+}
+
+module "secrets_manager" {
+  count = var.use_rds || var.use_s3 ? 1 : 0
+  source            = "../../../../iaac/terraform/common/aws-secrets-manager"
+  helm_config = {
+    chart = "${var.kf_helm_repo_path}/charts/common/aws-secrets-manager/rds-s3"
```
Comment (surajkota) [inline deployments/rds-s3/terraform/rds-s3-components/main.tf:88]: this also needs to change according to deployment option?
https://github.com/awslabs/kubeflow-manifests/tree/main/charts/common/aws-secrets-manager changes according to option

File: deployments/rds-s3/terraform/rds-s3-components/main.tf [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+}
+
+module "filter_kfp_set_values" {
+  source            = "../../../../iaac/terraform/utils/set-values-filter"
+  set_values = {
+    "rds.dbHost" = try(module.rds[0].rds_endpoint, null),
+    "s3.bucketName" = try(module.s3[0].s3_bucket_name ,null),
+    "s3.minioServiceRegion" = var.addon_context.aws_region_name,
+    "rds.mlmdDb" = var.mlmdb_name,
+    "s3.minioServiceHost" = var.minio_service_host
```
Comment (surajkota) [inline deployments/rds-s3/terraform/rds-s3-components/main.tf:185]: aws_region_name? shouldnt this be S3 endpoint?
also expose minioServiceHost and minioServiceRegion to the top level for people using private link and non commercial regions

File: deployments/rds-s3/terraform/rds-s3-components/variables.tf [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  description = "Name of the mlm DB to create"
+}
+
+variable "minio_service_host" {
+  type        = string
+  default = "s3.amazonaws.com"
+  description = "S3 service host DNS. This field will need to be changed when making requests from other partitions e.g. China Regions"
+}
+
+variable "secret_recovery_window_in_days" {
```
Comment (surajkota) [inline deployments/rds-s3/terraform/rds-s3-components/variables.tf:102]: is there an option for force destroy secret as well?

File: deployments/rds-s3/terraform/variables.tf [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  default = false
+}
+
+variable "force_destroy_s3_bucket" {
+  type = bool
+  description = "Destroys s3 bucket even when the bucket is not empty"
+  default = false
+}
+
+variable "aws_access_key" {
```
Comment (surajkota) [inline deployments/rds-s3/terraform/variables.tf:99]: change var names to MINIO_AWS_ACCESS_KEY_ID and MINIO_AWS_SECRET_ACCESS_KEY

File: iaac/terraform/aws-infra/rds/main.tf [hunk 6]
```diff
@@ -0,0 +1,45 @@
+resource "aws_db_subnet_group" "rds_db_subnet_group" {
```
Comment (surajkota) [inline iaac/terraform/aws-infra/rds/main.tf:1]: can you confirm that this configuration matching the one from auto-rds-s3 script?

File: iaac/terraform/aws-infra/rds/main.tf [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  subnet_ids = var.subnet_ids
+}
+
+resource "random_uuid" "db_snapshot_suffix" {
+  keepers = {
+    instance_class       = var.db_class
+    db_name                 = var.db_name
```
Comment (surajkota) [inline iaac/terraform/aws-infra/rds/main.tf:8]: nit: linting

File: iaac/terraform/aws-infra/rds/main.tf [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    multi_az = var.multi_az
+    db_subnet_group_name = aws_db_subnet_group.rds_db_subnet_group.id
+    security_group_id = var.security_group_id
+  }
+}
+
+resource "aws_db_instance" "kubeflow_db" {
+  allocated_storage    = var.db_allocated_storage
+  engine               = "mysql"
+  engine_version       = "8.0.30"
```
Comment (surajkota) [inline iaac/terraform/aws-infra/rds/main.tf:20]: can you please expose this as var as well?

File: iaac/terraform/aws-infra/s3/main.tf [hunk 9]
```diff
@@ -0,0 +1,17 @@
+resource "aws_s3_bucket" "artifact_store" {
```
Comment (surajkota) [inline iaac/terraform/aws-infra/s3/main.tf:1]: shall we apply the https only policy by default?

File: iaac/terraform/aws-infra/s3/variables.tf [hunk 10]
```diff
@@ -0,0 +1,20 @@
+variable "aws_access_key" {
+  type        = string
+  description = "AWS access key to authenticate minio client"
```
Comment (surajkota) [inline iaac/terraform/aws-infra/s3/variables.tf:3]: use the terms access key ID and secret access key everywhere

**Review by surajkota — CHANGES_REQUESTED** [review]
Reviewed at high level, looks good with few comments

File: deployments/rds-s3/terraform/Makefile [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+	terraform destroy -target="module.eks_blueprints" -auto-approve
+
+delete-eks-blueprints-k8s-addons:
+	terraform destroy -target="module.eks_blueprints_kubernetes_addons" -auto-approve
+
+delete-kubeflow-components:
+# hack to handle the automatic deletion of the sa when calling helm delete 
+# which causes deletion of the tf created sa to fail
+# todo: profiles-and-kfam helm chart should accept an optional irsa role annotation
+	terraform state rm "module.kubeflow_components.module.kubeflow_profiles_and_kfam.kubernetes_service_account_v1.profile_controller_sa" || true
```
[the remark this answers is not in the document]
Comment (rrrkharse) [reply deployments/rds-s3/terraform/Makefile:45]: Actually this is only removing the state of the k8s service account `kubernetes_service_account_v1.profile_controller_sa`

The module responsible for creating the role `module.kubeflow_components.module.kubeflow_profiles_and_kfam.module.irsa` is getting deleted

File: deployments/rds-s3/terraform/rds-s3-components/variables.tf (hunk 4 above)
Comment (rrrkharse) [reply deployments/rds-s3/terraform/rds-s3-components/variables.tf:102]: force destroy is set recover window to 0 days

File: iaac/terraform/aws-infra/s3/main.tf (hunk 9 above)
Comment (rrrkharse) [reply iaac/terraform/aws-infra/s3/main.tf:1]: will add a todo for now, dont want to test new configurations so i can focus on the cognito and cognito-rds-s3 PRs

File: iaac/terraform/aws-infra/rds/main.tf (hunk 6 above)
Comment (rrrkharse) [reply iaac/terraform/aws-infra/rds/main.tf:1]: https://github.com/awslabs/kubeflow-manifests/blob/main/tests/e2e/utils/rds-s3/auto-rds-s3-setup.py#L218-L229

File: iaac/terraform/aws-infra/rds/main.tf (hunk 7 above)
Comment (rrrkharse) [reply iaac/terraform/aws-infra/rds/main.tf:8]: fixed

File: iaac/terraform/aws-infra/rds/main.tf (hunk 6 above)
Comment (rrrkharse) [reply iaac/terraform/aws-infra/rds/main.tf:1]: https://github.com/awslabs/kubeflow-manifests/blob/main/tests/e2e/utils/rds-s3/auto-rds-s3-setup.py#L218-L229

File: iaac/terraform/aws-infra/s3/variables.tf (hunk 10 above)
Comment (rrrkharse) [reply iaac/terraform/aws-infra/s3/variables.tf]: done

File: iaac/terraform/aws-infra/rds/main.tf (hunk 8 above)
Comment (rrrkharse) [reply iaac/terraform/aws-infra/rds/main.tf]: done

File: iaac/terraform/aws-infra/rds/main.tf (hunk 7 above)
Comment (rrrkharse) [reply iaac/terraform/aws-infra/rds/main.tf]: fixed

File: deployments/rds-s3/terraform/variables.tf (hunk 5 above)
Comment (rrrkharse) [reply deployments/rds-s3/terraform/variables.tf]: done

File: deployments/rds-s3/terraform/rds-s3-components/main.tf (hunk 3 above)
Comment (rrrkharse) [reply deployments/rds-s3/terraform/rds-s3-components/main.tf:196]: > aws_region_name? shouldnt this be S3 endpoint?
s3.minioServiceRegion should be a region, s3.minioServiceHost is the endpoint

> also expose minioServiceHost and minioServiceRegion to the top level for people using private link and non commercial regions
done

File: deployments/rds-s3/terraform/rds-s3-components/main.tf (hunk 2 above)
Comment (rrrkharse) [reply deployments/rds-s3/terraform/rds-s3-components/main.tf]: fixed

File: deployments/rds-s3/terraform/rds-s3-components/main.tf (hunk 1 above)
Comment (rrrkharse) [reply deployments/rds-s3/terraform/rds-s3-components/main.tf]: fixed

File: deployments/rds-s3/terraform/main.tf [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+  secrets_store_csi_driver_helm_config = {
+    namespace   = "kube-system"
+    set = [
+      {
+        name = "syncSecret.enabled",
+        value = "true"
+      },
+      {
+        name = "enableSecretRotation",
```
[the remark this answers is not in the document]
Comment (rrrkharse) [reply deployments/rds-s3/terraform/main.tf]: removed

File: deployments/rds-s3/terraform/rds-s3-components/main.tf (hunk 3 above)
Comment (surajkota) [reply deployments/rds-s3/terraform/rds-s3-components/main.tf:196]: yep, you got my point

File: iaac/terraform/aws-infra/rds/main.tf (hunk 6 above)
Comment (rrrkharse) [reply iaac/terraform/aws-infra/rds/main.tf:1]: added the following options from the script
```
BackupRetentionPeriod=DB_BACKUP_RETENTION_PERIOD,
StorageType=DB_STORAGE_TYPE,
DeletionProtection=True,
MaxAllocatedStorage=DB_MAX_STORAGE,
PubliclyAccessible=False,
```

**Review by surajkota — APPROVED** [review]
Ship it!!

**Pull request merged**